sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/ApiIOError.java

```diff
@@ -30,12 +36,36 @@
 @AutoValue
 public abstract class ApiIOError {
 
+  static <T, ErrorT extends Exception> ApiIOError of(
+      @NonNull ErrorT e, @NonNull Coder<T> coder, @NonNull T element) throws CoderException {
+
+    String encodedElement = CoderUtils.encodeToBase64(coder, element);
+
+    return ApiIOError.builder()
+        .setEncodedElementAsBase64String(encodedElement)
+        .setMessage(Optional.ofNullable(e.getMessage()).orElse(""))
+        .setObservedTimestamp(Instant.now())
+        .setStackTrace(Throwables.getStackTraceAsString(e))
+        .build();
+  }
+
   static Builder builder() {
     return new AutoValue_ApiIOError.Builder();
   }
 
-  /** The encoded UTF-8 string representation of the related processed element. */
-  public abstract String getEncodedElementAsUtfString();
+  /**
+   * The Base64 string representation of the element encoded using {@link
+   * org.apache.beam.sdk.coders.Coder#encode}.
+   */
+  public abstract String getEncodedElementAsBase64String();
+
+  /**
+   * Decodes the Base64 string representation of {@link #getEncodedElementAsBase64String()} into
+   * original type using its {@link Coder}.
+   */
+  public <T> T decodeElement(Coder<T> coder) throws CoderException {
+    return CoderUtils.decodeFromBase64(coder, getEncodedElementAsBase64String());
```

Comment:
I think this will only work if the bytes encoded in the base64 format were generated using the same coder.

Reply:
I removed this and relied on a JSON string to associate the request with the error.